speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/SQLDatabaseStorage.cpp

```diff
@@ -1,229 +1,487 @@
-{
-    if (p_sqliteExec)
+    try
     {
-        sqlite3_close(p_db);
-        remove(p_databasePath.c_str());
-        if (p_zErrMsg != nullptr)
-        {
-            std::string s(p_zErrMsg);
-            p_error +=
-                "\n\tReturned with error: "
-                + s;
-        }
-        p_error += "\n";
-        sqlite3_free(p_zErrMsg);
-        p_inputFile.close();
-        throw std::exception(p_error.c_str());
+        int trial_id = InsertInitialData();
+        InsertSimulationData(trial_id);
     }
+    catch (std::exception& e) {throw e;}
+
+    m_inputFile.close();
 }
 
-/// @brief                  Creates SQL query that will create a database table
-/// @param p_inputFile      data file to read from
-/// @param p_reading        current data read from file
-/// @param p_tableName      name of table to create
-/// @param p_sqlStream      string stream that collects all sql statements
-/// @param p_headerTypes    map that sets the correct type for each header name
-/// @param p_variableAmount how many variables the table has
-/// @param p_headers        header names
-void CreateTableSQL(
-        std::ifstream& p_inputFile,
-        std::string& p_reading,
-        const std::string& p_tableName,
-        std::stringstream& p_sqlStream,
-        const std::unordered_map<std::string,std::string>& p_headerTypes,
-        int& p_variableAmount,
-        std::string& p_headers)
+/// @brief Connect to the specified database. Initialise the database with the proper tables if they don't exist yet.
+/// @param p_hostName Hostname of the database to connect to. Should be a TCP-IP address.
+/// @param p_port Port the database is located on on the host.
+/// @param p_username Username to connect with to the database.
+/// @param p_password Password to connect with to the database.
+/// @param p_schemaName Name of the database schema to use.
+/// @return returns true if connection to database has been made, false otherwise
+bool SQLDatabaseStorage::OpenDatabase(
+    const std::string& p_hostName,
+    int p_port,
+    const std::string& p_username,
+    const std::string& p_password,
+    const std::string& p_schemaName)
 {
-    std::stringstream headersStream;
+    // Initialise SQL driver
+    m_driver = sql::mysql::get_mysql_driver_instance();
 
-    p_sqlStream.clear();
-    p_sqlStream.str("");
+    // Set connection options, and connect to the database
+    sql::ConnectOptionsMap connection_properties;
+    connection_properties["hostName"] = "tcp://" + p_hostName;
+    connection_properties["userName"] = p_username;
+    connection_properties["password"] = p_password;
+    connection_properties["port"] = p_port;
+    connection_properties["OPT_RECONNECT"] = true;
+    connection_properties["CLIENT_MULTI_STATEMENTS"] = false;
+    try { m_connection = m_driver->connect(connection_properties);}
```

Comment:
does this actually throw an exception, or can you check the connection with a function of m_connection, or can you check it using if(m_connection)?

Reply:
It does throw an exception, just tested it again to confirm